### PR TITLE
Crimson/osd:  add error log

### PR DESCRIPTION
**Pull request by liu-chunmei**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/crimson/osd/osd_operations/logmissing_request.cc [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    return crimson::get_logger(ceph_subsys_osd);
+  }
+}
+
+namespace crimson::osd {
+
+LogMissingRequest::LogMissingRequest(crimson::net::ConnectionRef&& conn,
+		       Ref<MOSDPGUpdateLogMissing> &&req)
+  : conn{std::move(conn)},
+    req{req}
```
Comment (athanatos) [inline src/crimson/osd/osd_operations/logmissing_request.cc:24]: req{std::move{req}} ?

File: src/crimson/osd/osd_operations/logmissing_request_reply.cc [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  }
+}
+
+namespace crimson::osd {
+
+LogMissingRequestReply::LogMissingRequestReply(
+  crimson::net::ConnectionRef&& conn,
+  Ref<MOSDPGUpdateLogMissingReply> &&req)
+  : conn{std::move(conn)},
+    req{req}
```
Comment (athanatos) [inline src/crimson/osd/osd_operations/logmissing_request_reply.cc:25]: same here

File: src/crimson/osd/shard_services.h [hunk 3]
```diff
@@ -139,6 +139,13 @@ class ShardServices : public md_config_obs_t {
     return dispatch_context({}, std::move(ctx));
   }
 
+  // -- tids --
+  // for ops i issue
+  std::atomic<unsigned int> last_tid{0};
```
Comment (athanatos) [inline src/crimson/osd/shard_services.h:144]: No need for std::atomic here.

File: src/crimson/osd/replicated_backend.cc [hunk 4]
```diff
@@ -53,7 +53,7 @@ ReplicatedBackend::_submit_transaction(std::set<pg_shard_t>&& pg_shards,
     throw crimson::common::actingset_changed(peering->is_primary);
   }
 
-  const ceph_tid_t tid = next_txn_id++;
+  const ceph_tid_t tid = shard_services.get_tid();
```
Comment (athanatos) [inline src/crimson/osd/replicated_backend.cc:56]: Is this because the tid needs to be unique across pgs?

**Comment by athanatos** [comment]
Squash the last two commits.

File: src/crimson/osd/pg.cc [hunk 5]
```diff
@@ -763,18 +763,107 @@ PG::do_osd_ops(
...
+          }
+          reply->set_enoent_reply_versions(
+          peering_state.get_info().last_update,
+          peering_state.get_info().last_user_version);
+          reply->add_flags(CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK);
+          return do_osd_ops_iertr::make_ready_future<MURef<MOSDOpReply>>(
+	    std::move(reply));
+	};
+
+      if(!peering_state.pg_has_reset_since(epoch) && op_info.may_write()) {
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:841]: Formatting

File: src/crimson/osd/pg.cc [hunk 6]
```diff
@@ -763,18 +763,107 @@ PG::do_osd_ops(
...
+	mempool::osd_pglog::list<pg_log_entry_t> log_entries;
+	log_entries.push_back(pg_log_entry_t(pg_log_entry_t::ERROR,
+	                                     obc->obs.oi.soid,
+                                             next_version(),
+					     eversion_t(), 0,
+	                                     reqid, utime_t(),
+	                                     -e.value()));
+	if (op_info.allows_returnvec()) {
+	  log_entries.back().set_op_returns(m->ops);
+	}
```
Comment (athanatos) [inline src/crimson/osd/pg.cc:784]: Let's factor out a submit_error_log helper here.

File: src/crimson/osd/replicated_backend.cc (hunk 4 above)
Comment (liu-chunmei) [reply src/crimson/osd/replicated_backend.cc:56]: yes.

Comment (athanatos) [reply src/crimson/osd/replicated_backend.cc:56]: This isn't actually correct.  The tid on a repop should match the tid from the original client request.  PrimaryLogPG::do_op constructs the OpContext for the op using MOSDOp::get_reqid():

```

  dout(25) << __func__ << " oi " << obc->obs.oi << dendl;

  OpContext *ctx = new OpContext(op, m->get_reqid(), &m->ops, obc, this);

  if (m->has_flag(CEPH_OSD_FLAG_SKIPRWLOCKS)) {
    dout(20) << __func__ << ": skipping rw locks" << dendl;
  } else if (m->get_flags() & CEPH_OSD_FLAG_FLUSH) {
    dout(20) << __func__ << ": part of flush, will ignore write lock" << dendl;
```

This gets assigned to OpContext::reqid which is then used to populate the rep_tid field on the repop:

```

  // issue replica writes
  ceph_tid_t rep_tid = osd->get_tid();

  RepGather *repop = new_repop(ctx, rep_tid);

  issue_repop(repop, ctx);
  eval_repop(repop);
```
PrimaryLogPG::issue_repop then passes repop->rep_tid as the tid:

```
 recovery_state.pre_submit_op(
    soid,
    ctx->log,
    ctx->at_version);
  pgbackend->submit_transaction(
    soid,
    ctx->delta_stats,
    ctx->at_version,
    std::move(ctx->op_t),
    recovery_state.get_pg_trim_to(),
    recovery_state.get_min_last_complete_ondisk(),
    std::move(ctx->log),
    ctx->updated_hset_history,
    on_all_commit,
    repop->rep_tid,
    ctx->reqid,
    ctx->op);
}
```

ReplicatedBackend::generate_subop then uses that tid to populate the MOSDRepOp:

```
Message * ReplicatedBackend::generate_subop(
  const hobject_t &soid,
  const eversion_t &at_version,
  ceph_tid_t tid,
  osd_reqid_t reqid,
  eversion_t pg_trim_to,
  eversion_t min_last_complete_ondisk,
  hobject_t new_temp_oid,
  hobject_t discard_temp_oid,
  const bufferlist &log_entries,
  std::optional<pg_hit_set_history_t> &hset_hist,
  ObjectStore::Transaction &op_t,
  pg_shard_t peer,
  const pg_info_t &pinfo)
{
  int acks_wanted = CEPH_OSD_FLAG_ACK | CEPH_OSD_FLAG_ONDISK;
  // forward the write/update/whatever
  MOSDRepOp *wr = new MOSDRepOp(
    reqid, parent->whoami_shard(),
    spg_t(get_info().pgid.pgid, peer.shard),
    soid, acks_wanted,
    get_osdmap_epoch(),
    parent->get_last_peering_reset_epoch(),
    tid, at_version);
```

PrimaryLogPG::record_write_error and PrimaryLogPG::submit_log_entries actually seem to use PrimaryLogPG::new_repop which uses osd->get_tid() rather than the tid on the client op.  I think this is actually a bug -- submit_log_entries predates error logging and the error logging pathway should really be using the client operation's tid.  @jdurgin Do you recall whether there's a good reason to be using a fresh tid for those operations?

Users of simple_opc_submit do need to generate their own OSD-global tid's since they are generated internally to the osd (watch expiration, snap trim, etc) rather than on behalf of a client operation.  The only similar user crimson has at the moment would be watch expiration, but I don't see the logic for that yet.  @rzarzynski Where's the watch expiration logic?

In summary, normal client IO and error log entries should use the tid from the original MOSDOp.

**Review by athanatos — CHANGES_REQUESTED** [review]

File: src/crimson/osd/replicated_backend.cc (hunk 4 above)
Comment (jdurgin) [reply src/crimson/osd/replicated_backend.cc:56]: It's a bug, it was not intended to use a new tid for error log entries.

Comment (liu-chunmei) [reply src/crimson/osd/replicated_backend.cc:56]: @athanatos there are two data structure, one is ceph_tid_t ( often named tid) another is osd_reqid_t (often named reqid which contains client tid), in classic osd, they use osd->get_tid() to create its own tid, and use OpContext->reqid (get from client request by m->get_reqid()) as reqid when create MOSDRepOp.  So it seems not a bug,  two id one stands for osd, one stands for client.  What do you think?

Comment (liu-chunmei) [reply src/crimson/osd/replicated_backend.cc:56]: MOSDRepOp *wr = new MOSDRepOp(
    reqid, parent->whoami_shard(),   // here reqid is get from OpContext->reqid get from m->get_reqid()
    spg_t(get_info().pgid.pgid, peer.shard),
    soid, acks_wanted,
    get_osdmap_epoch(),
    parent->get_last_peering_reset_epoch(),
    tid, at_version);    // here tid is ceph_tid_t get from osd->get_tid();

so seems reqid and tid shouldn't contain the same infomation.

Comment (athanatos) [reply src/crimson/osd/replicated_backend.cc:56]: You are right!  I'm sorry, I managed to totally confuse myself.

**Review by athanatos — APPROVED** [review]

**Comment by athanatos** [comment]
jenkins test make check
